okhttp/src/main/java/okhttp3/Address.kt

```diff
@@ -0,0 +1,156 @@
+   * this URL are always empty, since they are not significant for planning a route.
+   */
+  fun url() = url
+
+  /** Returns the service that will be used to resolve IP addresses for hostnames. */
+  fun dns() = dns
+
+  /** Returns the socket factory for new connections. */
+  fun socketFactory(): SocketFactory {
+    return socketFactory
+  }
+
+  /** Returns the client's proxy authenticator. */
+  fun proxyAuthenticator() = proxyAuthenticator
+
+  /**
+   * Returns the protocols the client supports. This method always returns a non-null list that
+   * contains minimally {@link Protocol#HTTP_1_1}.
+   */
+  fun protocols() = protocols
+
+  fun connectionSpecs() = connectionSpecs
+
+  /**
+   * Returns this address's proxy selector. Only used if the proxy is null. If none of this
+   * selector's proxies are reachable, a direct connection will be attempted.
+   */
+  fun proxySelector() = proxySelector
+
+  /**
+   * Returns this address's explicitly-specified HTTP proxy, or null to delegate to the {@linkplain
+   * #proxySelector proxy selector}.
+   */
+  fun proxy() = proxy
+
+  /** Returns the SSL socket factory, or null if this is not an HTTPS address. */
+  fun sslSocketFactory() = sslSocketFactory
+
+  /** Returns the hostname verifier, or null if this is not an HTTPS address. */
+  fun hostnameVerifier() = hostnameVerifier
+
+  /** Returns this address's certificate pinner, or null if this is not an HTTPS address. */
+  fun certificatePinner() = certificatePinner
+
+  override fun equals(other: Any?): Boolean {
+    return (other is Address
+        && url == other.url
+        && equalsNonHost((other as Address?)!!))
+  }
+
+  override fun hashCode(): Int {
+    var result = 17
+    result = 31 * result + url.hashCode()
+    result = 31 * result + dns.hashCode()
+    result = 31 * result + proxyAuthenticator.hashCode()
+    result = 31 * result + protocols.hashCode()
+    result = 31 * result + connectionSpecs.hashCode()
+    result = 31 * result + proxySelector.hashCode()
+    result = 31 * result + Objects.hashCode(proxy)
+    result = 31 * result + Objects.hashCode(sslSocketFactory)
+    result = 31 * result + Objects.hashCode(hostnameVerifier)
+    result = 31 * result + Objects.hashCode(certificatePinner)
+    return result
+  }
+
+  internal fun equalsNonHost(that: Address): Boolean {
+    return (this.dns == that.dns
+        && this.proxyAuthenticator == that.proxyAuthenticator
+        && this.protocols == that.protocols
+        && this.connectionSpecs == that.connectionSpecs
+        && this.proxySelector == that.proxySelector
+        && this.proxy == that.proxy
+        && this.sslSocketFactory == that.sslSocketFactory
+        && this.hostnameVerifier == that.hostnameVerifier
+        && this.certificatePinner == that.certificatePinner
+        && this.url().port() == that.url().port())
+  }
+
+  override fun toString(): String {
+    val result = StringBuilder()
```

Comment:
buildString ?

Reply:
Done in e751fc152f6b74804a016a119e8fbc64ae85142f